Updating the UI

skylerknecht opened the pull request:
As per, https://github.com/skylerknecht/messenger/issues/6, we need to revamp the UI.

I pushed a couple changes to this branch to solve the ToDos listed in the issue. 

Overall, Messenger IDs didn't match and the help menus were inaccurate. I've updated the UI
to permit operators to interact with Messengers with their ID instead of providing the ID
in commands. 

Please walkthrough a full workflow with the new UI and let me know if there should be any
additional changes to the UI.

Thanks!

Invoke-Mimikatz commented:
Nothing in the help menu indicates that a messenger can be interacted with.

```
Available commands:
  exit       Exit the application, stopping the messenger server.
  back       Returns to the main menu.
  forwarders Display a list of active forwarders in a table format.
  messengers Display a list of messengers in a table format.
  local      Start a local forwarder for the specified messenger.
  remote     Start a remote forwarder.
  socks      Start a socks proxy.
  stop       Stop a forwarder.
  help       Print a list of commands and their descriptions.
  ?          Print a list of commands and their descriptions.
  ```
You should have an `interact` command, even if you also allow messengers to be interacted with based on simply inputting the messenger ID.
These two should be equivalent:
```
UkGLdgnCfz
interact UkGLdgnCfz
```

When displaying forwarders, make sure to indicate if no matching forwarders exist for a specific messenger:
```
(messenger)~# forwarders UkGLdgnCfz
                                      Forwarders
Type Identifier Clients Listening Host Listening Port Destination Host Destination Port
---- ---------- ------- -------------- -------------- ---------------- ----------------

```
This should instead say something like,
```
[*] There are currently no forwarders for messenger 'UkGLdgnCfz'
```
Or if a non-existant messenger is specified,
```
[*] Cannot show forwarders for non-existant messenger id: 'asdfDoesntExist'
```
Additionally, update the forwarders example help to use the new messenger ID scheme:
```
examples:
  forwarders
  forwarders 4369562880
  ```
  Should be:
  ```
  examples:
  forwarders
  forwarders UkGLdgnCfz
  ```
  
  If there are no forwarders at all, simply list that there are no current forwarders:
  ```
  [*] There are currently no forwarders to display!
  ```
  
  If socks, local, and remote forwarders can only be set up after interacting with a messenger (which I like this idea), the commands should only be available when interacting with a messenger. Consider breaking up the help menu for commands into server commands and messenger commands:
  ```
  Available server commands:
    forwarders Display a list of active forwarders in a table format.
    messengers Display a list of messengers in a table format.
    stop       Stop a forwarder.
    exit       Exit the application, stopping the messenger server.
    back       Returns to the main menu.
    help       Print a list of commands and their descriptions.
  
  Messenger commands (must be interacting with a messenger):
    local      Start a local forwarder for the specified messenger.
    remote     Start a remote forwarder.
    socks      Start a socks proxy.
  ```
  
  Consider bolding (strong, white text) the messenger ID (`UkGLdgnCfz`), since this is an important value and implies to the operator that this value is significant throughout the messenger application.
  
  Consider adding bash-like double tab completion to display all options instead of cycling through options.
  
  Consider adding a link in the help to a more detailed usage guide with examples and common messenger use-cases.
  - Standard SOCKS proxy
  - Reverse port forward a web server
  - Chaining messenger clients with multiple reverse port forwards
  - NTLMRelay2self setup
  - Details on protocol implementation, performance, operational nuances

skylerknecht commented:
Excellent, thanks for the review. Here's the current tracker:

- [x] Updated the Messenger IDs to match the new help menus. 
- [x] Added a message to notify users when there are no forwarders or messengers. 
- [x] Added a interact command and associated docs.
- [ ] Add a message to notify users that the Messenger ID they're trying to use does not exist.
- [ ] Add a link to associated TTPs that can be used with SOCKs proxies, Local Port Forwards and Remote Port Forwards.
- [ ] Bold important text like the Messenger IDs of the Messenger you're currently interacting with.

Invoke-Mimikatz reviewed (approved):
LGTM MERGE IT
